### DISCUSSION TITLE
"Call from invalid thread" exception in Brush.Parse();

### DISCUSSION THREAD

**Discussion by Mochengvia** [AvaloniaUI/Avalonia#6376] [Q&A] [answered]
I placed a Brush.Parse() method in IValueConverter and used the converter in my DataGridTemplateColumn binding. 
About the fifth time the Convert method is called, this error will be thrown. 

![image](https://user-images.githubusercontent.com/23360265/128474884-0287775b-b8a0-4de3-9512-8f79921b7a3c.png)

Here is the full StackTrace:

```
"   at Avalonia.Threading.Dispatcher.VerifyAccess() in /_/src/Avalonia.Base/Threading/Dispatcher.cs:line 48\r\n   at Avalonia.Media.Brush..ctor()\r\n   at Avalonia.Media.SolidColorBrush..ctor(Color color, Double opacity) in /_/src/Avalonia.Visuals/Media/SolidColorBrush.cs:line 37\r\n   at Avalonia.Media.Brush.Parse(String s) in /_/src/Avalonia.Visuals/Media/Brush.cs:line 50\r\n   at Fenikso.Frameworks.Converters.RiseFallForegroundConverter.Convert(Object value, Type targetType, Object parameter, CultureInfo culture) in E:\\Fenikso\\SharedResources\\Fenikso.Frameworks\\Converters\\RiseFallForegroundConverter.cs:line 12\r\n   at Avalonia.Data.Core.BindingExpression.ConvertValue(Object value) in /_/src/Avalonia.Base/Data/Core/BindingExpression.cs:line 215\r\n   at Avalonia.Data.Core.BindingExpression.InnerListener.OnNext(Object value) in /_/src/Avalonia.Base/Data/Core/BindingExpression.cs:line 362\r\n   at Avalonia.Reactive.LightweightObservableBase`1.PublishNext(T value) in /_/src/Avalonia.Base/Reactive/LightweightObservableBase.cs:line 148\r\n   at Avalonia.Data.Core.ExpressionObserver.ValueChanged(Object value) in /_/src/Avalonia.Base/Data/Core/ExpressionObserver.cs:line 315\r\n   at Avalonia.Data.Core.ExpressionNode.ValueChanged(Object value, Boolean notify) in /_/src/Avalonia.Base/Data/Core/ExpressionNode.cs:line 127\r\n   at Avalonia.Data.Core.ExpressionNode.ValueChanged(Object value) in /_/src/Avalonia.Base/Data/Core/ExpressionNode.cs:line 94\r\n   at Avalonia.Data.Core.Plugins.PropertyAccessorBase.PublishValue(Object value) in /_/src/Avalonia.Base/Data/Core/Plugins/PropertyAccessorBase.cs:line 61\r\n   at Avalonia.Data.Core.Plugins.InpcPropertyAccessorPlugin.Accessor.SendCurrentValue() in /_/src/Avalonia.Base/Data/Core/Plugins/InpcPropertyAccessorPlugin.cs:line 171"
```

**Comment by jmacato** [comment]
seems like the converter is being called from a non UI thread. try enclosing that convert method in to a Dispatcher like this:

```
			Dispatcher.UIThread.InvokeAsync(() =>
			{
				// Code here;
			});
```

**Reply by Mochengvia** [reply]
I think you are joking, I need to get a return value immediately instead of using Async here

**Reply by Mochengvia** [reply]
This is an IValueConverter, not a business logic or something

**Reply by jmacato** [reply]
Please be  professional with your tone. I'm not joking and i'm trying to help you here.

You can do something like this:

```

Brush outBrush;

Dispatcher.UIThread.Post(() =>
{ 
     outBrush = Brush.Parse("White");
});


```

**Reply by Mochengvia** [reply]
Unfortunately, I tried this method before asking this question. 
The content in Dispatcher.UIThread.Post will not be called immediately. 

![image](https://user-images.githubusercontent.com/23360265/128477397-b49f5d97-6cff-47b1-bdaa-e4559f377daf.png)

**Reply by jmacato** [reply]
hmm.. i wonder if the value you're setting in business logic can be called in dispatcher instead? 

Though honestly we need an Immutable brush parser so this wont happen much.

**Reply by Mochengvia** [reply]
The order of hitting breakpoints is 1, 3, and after a long period of time, breakpoint 2 will be hit.

**Reply by Mochengvia** [reply]
You mean like this? It does not work.

![image](https://user-images.githubusercontent.com/23360265/128478397-fd5f4619-d473-42e0-89ca-2995d339f8c8.png)

I have raised several bugs for Avalonia, I just can’t understand why you locked the Issue without check.

**Reply by Mochengvia** [reply]
The second breakpoint will be hit a long time after return.

**Reply by MihaMarkic** [reply]
You have to use dispatcher sooner, when you set the property that is bound. I'd never use threading in a converter or UI in general.

**Reply by Mochengvia** [reply]
I'm not sure what you mean...this is where I use Converter. 
Is this the wrong place to use? 

![image](https://user-images.githubusercontent.com/23360265/128479930-587ed351-b891-4dc7-a1ba-21560266342d.png)

All I did was return my custom brushes in the Converter. Is there any way to achieve this?

**Reply by FoggyFinder** [reply]
can you provide MCVE?

**Reply by Mochengvia** [reply]
Alright, I'm on it.

**Reply by MihaMarkic** [reply]
@Mochengvia When you assign the value to IncreaseRatio. The assignment should be in UI thread.

**Comment by kekekeks** [comment, marked as answer]
It looks like you are changing a property of your view model on non-UI thread. This is not supported, all property change notifications should happen on the UI thread.

**Reply by Mochengvia** [reply]
I used my Converter here. Does the use here belong to non-UI Thread? 

https://github.com/AvaloniaUI/Avalonia/discussions/6376#discussioncomment-1138112

**Reply by kekekeks** [reply]
`IncreaseRatio` is somehow changed on non-UI thread. 

If you are using ReactiveUI, that might be caused by missing initialization logic or your container registration replacing our dispatcher. Misconfigured ReactiveUI uses thread pool for its observables.

**Reply by danwalmsley** [reply]
Do you have UseReactiveUI() call on your AppBuilder?

**Reply by Mochengvia** [reply]
No, I am not using ReactiveUI

**Reply by Gillibald** [reply]
Brush.Parse produces immutable brushes in current master. So this might not be an issue with a nightly build.

**Comment by Mochengvia** [comment]
Anyway, thank you all for your help.

I restore here the reason why Dispatcher.UIThread.Post cannot be used in this situation, but I cannot reproduce the error "Call from invalid thread" in the new project. 

Here is the source code:

https://mega.nz/file/D1ZVjSxY#Ia5xWl6eN25VIUyyVgg8L3y34b7hkr0od_uHGn5VxIc

I am still not sure what caused the cross-thread error. I created Converter in UserControl.Resources and used it in DataGridTemplateColumn. 
When the Convert method is called, I can see the first few Brush.Parse() method executed correctly, but this exception must be thrown when the breakpoint is hit for the fifth time. 

![image](https://user-images.githubusercontent.com/23360265/128484863-c0365be8-3c06-4566-b2eb-35984a67afc4.png)
![image](https://user-images.githubusercontent.com/23360265/128484933-29c85482-4f4b-44fb-b38f-4faacb4885a9.png)

This is an enterprise project, so I'm sorry I can't share it... We have made many packages on Avalonia, such as MVVM framework and lazy loading DataGrid. I am pretty sure that there is no cross-thread operation. I now use static Brushes.Red instead of the Brush.Parse() method, it can continue to work anyway. If you need me to test any code, I will continue to provide results here.

Also, I am sorry that my English has caused you troubles, because it comes from Google Translate

**Reply by kekekeks** [reply]
>I am still not sure what caused the cross-thread error. I created Converter in UserControl.Resources and used it in DataGridTemplateColumn.

The cross-thread error is not related to the converter. The property you are binding to got changed outside of the UI thread. Add `Dispatcher.UIThread.VerifyAccess()` to you property setter, so you'll get the exception when it got changed.

**Reply by Mochengvia** [reply]
You reminded me, I double-checked the code, this is indeed a problem caused by asynchrony. This is my mistake, sorry.
